File: app/src/main/java/com/przemyslawlusnia/vocabularycreator/core/RxUseCase.java
package com.przemyslawlusnia.vocabularycreator.core;

import rx.Observable;
import rx.Scheduler;

public abstract class RxUseCase<T> {

  private final Scheduler executionScheduler;

  public RxUseCase(Scheduler executionScheduler) {
    this.executionScheduler = executionScheduler;
  }

  protected abstract Observable<T> buildUseCaseObservable();

  public Observable<T> execute() {
    return buildUseCaseObservable().subscribeOn(executionScheduler);
  }
}

File: app/src/main/java/com/przemyslawlusnia/vocabularycreator/core/RxUseCase.kt
package com.przemyslawlusnia.vocabularycreator.core

import rx.Observable
import rx.Scheduler

abstract class RxUseCase<T>(private val executionScheduler: Scheduler) {

    protected abstract fun buildUseCaseObservable(): Observable<T>

    fun execute(): Observable<T> {
        return buildUseCaseObservable().subscribeOn(executionScheduler)
    }
}
